pkg/workflows/ring/plugin.go

```diff
@@ -0,0 +1,282 @@
+		WantShards:  wantShards,
+	}
+
+	return proto.MarshalOptions{Deterministic: true}.Marshal(observation)
+}
+
+//coverage:ignore
+func (p *Plugin) ValidateObservation(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, _ types.AttributedObservation) error {
+	return nil
+}
+
+func (p *Plugin) ObservationQuorum(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (quorumReached bool, err error) {
+	return quorumhelper.ObservationCountReachesObservationQuorum(quorumhelper.QuorumTwoFPlusOne, p.config.N, p.config.F, aos), nil
+}
+
+func (p *Plugin) collectShardInfo(aos []types.AttributedObservation) (shardHealth map[uint32]int, workflows []string, timestamps []time.Time, wantShardVotes []uint32) {
+	shardHealth = make(map[uint32]int)
+	for _, ao := range aos {
+		observation := &pb.Observation{}
+		if err := proto.Unmarshal(ao.Observation, observation); err != nil {
+			p.lggr.Warnf("failed to unmarshal observation: %v", err)
+			continue
+		}
+
+		for shardID, status := range observation.ShardStatus {
+			if status != nil && status.IsHealthy {
+				shardHealth[shardID]++
+			}
+		}
+
+		workflows = append(workflows, observation.WorkflowIds...)
+
+		if observation.Now != nil {
+			timestamps = append(timestamps, observation.Now.AsTime())
+		}
+
+		if observation.WantShards > 0 {
+			wantShardVotes = append(wantShardVotes, observation.WantShards)
+		}
+	}
+	return shardHealth, workflows, timestamps, wantShardVotes
+}
+
+func (p *Plugin) getHealthyShards(shardHealth map[uint32]int) []uint32 {
+	var healthyShards []uint32
+	for shardID, votes := range shardHealth {
+		if votes > p.config.F {
+			healthyShards = append(healthyShards, shardID)
+			p.store.SetShardHealth(shardID, true)
+		}
+	}
+	slices.Sort(healthyShards)
+
+	return healthyShards
+}
+
+func (p *Plugin) Outcome(_ context.Context, outctx ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	// Bootstrap with 1 shard on first round; subsequent rounds build on prior outcome
+	prior := &pb.Outcome{}
+	if outctx.PreviousOutcome == nil {
+		prior.Routes = map[string]*pb.WorkflowRoute{}
+		prior.State = &pb.RoutingState{Id: outctx.SeqNr, State: &pb.RoutingState_RoutableShards{RoutableShards: 1}}
+	} else if err := proto.Unmarshal(outctx.PreviousOutcome, prior); err != nil {
+		return nil, err
+	}
+
+	currentShardHealth, allWorkflows, nows, wantShardVotes := p.collectShardInfo(aos)
+
+	// Need at least F+1 timestamps; fewer means >F faulty nodes and we can't trust this round
+	if len(nows) < p.config.F+1 {
+		return nil, errors.New("insufficient observation timestamps")
+	}
+	slices.SortFunc(nows, time.Time.Compare)
+
+	// Use the median timestamp to determine the current time
+	now := nows[len(nows)/2]
+
+	// Use median for wantShards consensus; fall back to current state if insufficient votes
+	var wantShards uint32
+	if len(wantShardVotes) >= p.config.F+1 {
```

Comment:
This should always be the case if you validate observations like I suggested below. Then maybe you don't need all other else blocks?